src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -0,0 +1,1845 @@
+                //         object o = (System.Delegate)p.F;
+                Diagnostic(ErrorCode.ERR_NoExplicitConv, "(System.Delegate)p.F").WithArguments("method", "System.Delegate").WithLocation(8, 20));
+        }
+
+        [Fact]
+        public void MethodGroup_Inaccessible()
+        {
+            var source =
+@"using System;
+class A
+{
+    private static void F() { }
+    internal static void F(object o) { }
+}
+class B
+{
+    static void Main()
+    {
+        Delegate d = A.F;
+        Console.WriteLine(d.GetDelegateTypeName());
+    }
+}";
+            CompileAndVerify(new[] { source, s_utils }, parseOptions: TestOptions.RegularPreview, expectedOutput: "System.Action<System.Object>");
+        }
+
+        [Fact]
+        public void MethodGroup_IncorrectArity()
+        {
+            var source =
+@"class Program
+{
+    static void F0(object o) { }
+    static void F0<T>(object o) { }
+    static void F1(object o) { }
+    static void F1<T, U>(object o) { }
+    static void F2<T>(object o) { }
+    static void F2<T, U>(object o) { }
+    static void Main()
+    {
+        System.Delegate d;
+        d = F0<int, object>;
+        d = F1<int>;
+        d = F2;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (12,13): error CS0308: The non-generic method 'Program.F0(object)' cannot be used with type arguments
+                //         d = F0<int, object>;
+                Diagnostic(ErrorCode.ERR_HasNoTypeVars, "F0<int, object>").WithArguments("Program.F0(object)", "method").WithLocation(12, 13),
+                // (13,13): error CS0308: The non-generic method 'Program.F1(object)' cannot be used with type arguments
+                //         d = F1<int>;
+                Diagnostic(ErrorCode.ERR_HasNoTypeVars, "F1<int>").WithArguments("Program.F1(object)", "method").WithLocation(13, 13),
+                // (14,13): error CS8915: The delegate type could not be inferred.
+                //         d = F2;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F2").WithLocation(14, 13));
+        }
+
+        [Fact]
+        public void ExtensionMethods_01()
+        {
+            var source =
+@"static class E
+{
+    internal static void F1(this object x, int y) { }
+    internal static void F2(this object x) { }
+}
+class Program
+{
+    void F2(int x) { }
+    static void Main()
+    {
+        System.Delegate d;
+        var p = new Program();
+        d = p.F1;
+        d = p.F2;
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (14,15): error CS8915: The delegate type could not be inferred.
+                //         d = p.F2;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F2").WithLocation(14, 15));
```

Comment:
This isn't making sense to me: why can't this be inferred to `Action`?

Reply:
> why can't this be inferred to `Action`?

Because there are two overloads with distinct signatures: `Program.F2(int x)` and `E.F2(this object x)`.

Reply:
Ah, ok, thanks. I missed that one.

---
In reply to: [618776762](https://github.com/dotnet/roslyn/pull/52448#discussion_r618776762) [](ancestors = 618776762)